### ISSUE TITLE
Optional discriminated union's callback parameter type  is not inferred

### ISSUE THREAD

**Issue by giladsegal** [microsoft/TypeScript#35769]
**discriminated union**

**Code**

```ts
type T1 = {
        as?: 'button',
        onClick?: (el: number) => void
} 

type T2 = {
        as: 'a',
        onClick?: (el: string) => void
}

type T = T1 | T2;

// onClick first parameter type is inferred according to T1 (number)
const myT1_1: T = {
        as: undefined,
        onClick: _ev => {}
}

// onClick first parameter type is inferred according to T1 (number)
const myT1_2: T = {
        as: 'button',
        onClick: _ev => {}
}

// onClick first parameter type is inferred according to T2 (string)
const myT2_1: T = {
        as: 'a',
        onClick: _ev => {}
}

// onClick first parameter type is not inferred (any)
const myT1_3: T = {
        onClick: _ev => {}
} 
```

**Expected behavior:**
onClick `_ev` parameter of `myT1_3` is inferred to be a `number`

**Actual behavior:**
onClick `_ev` parameter of `myT1_3` is `any`

**Playground Link:** 
[Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=35&pc=2#code/C4TwDgpgBAKgjFAvFA3gWAFBWz7BDAZwH4AuKAcgCMBXYYAewDtyAaTXXJgYQBsBLAMYBrUlAAUEHmUbUAtpQgAnAJRIAfFABu9PgBNMAXyiZMoSLABMSVOw74CZcnla273fsNESpUAsEV8jADmqoga2nqGJhhm0DDW8FAAPpYA3NEA9BlQ7oJCUABmfIp+UGB4iniyEMBKULFQfASNjAVKihC6UHgCAvSKuoFB9fSwCGIy8krKmH2MpbIg8AD6cGTxyOhYdt0OUNSMuhBFjJ1s226MvHlkyxCa6qgGURiYWTlXHvlFJcBlFVUanUGk0Wm1FB0uj0+gMhiMxuJJgoVLMmAslnBlhZ1tYtjtdo4aHQmC4LhxcsJbvdHihnhg6W9shTvsVSuVKtVaop6uBoKDAuDId1ev1BsF4TArGI-AFgjMMHN0ZLVjjNq4OIRHM5zvjmVSHmEni9GR9rsJCqy-uzAVyeeZQYx6H8Be1OuI8IwQPLFX9FisAMyqmxkzifG5QO4GjS0wxAA)

If the callback in `T1` and `T2` is replaced with a primitive member then `myT1_3` is inferred correctly. [Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=34&pc=2#code/C4TwDgpgBAKgjFAvFA3gWAFBWz7BDAZwH4AuKAcgCMBXYYAewDtyAaTXXMAJ3rFKkbUAtpQhc2GAL5RMmUJFgAmJKnYd8BMuTys167r34FgXAJaMA5pkmyM86DBXwoAHyUBuWwHovUA2Ch7KFMCYMYAMzEuCAATKDwAYwT6LhjzC0D6WAQACkERMQBKTGTGYyghEHgAfTgyR2R0LHV4zShqRhiIcPNYiRa-HjAyOFsbDEwfQd5A8GgQsMiuaLjE5NT0zOyoPOFRLmKMUvLKmsV6lSaBwi0aOiZdZv0hketvX39ZhQXzJZX4pIpNKWLYwZQ5YxmSyHY7ACpVRS1C6NPQcG4UHT9Fr+LSKchvCYYKafII-CJRWIA9bAjIMba7AoHEpME5VODVADMyNUTw4OKgcAkkiAA)

**Comment by wvanvugt-speedline** [comment]
This appears to be a duplicate of https://github.com/microsoft/TypeScript/issues/41759

**Comment by IanVS** [comment]
The parameter seems to be correctly inferred in the latest version of typescript, can this be closed as fixed?  

![image](https://user-images.githubusercontent.com/4616705/120839894-83cc2780-c537-11eb-944d-f8c04f90fe05.png)

**Comment by RyanCavanaugh** [comment]
Confirmed fixed 4.2 -> 4.3

**Comment by wingsico** [comment]
@RyanCavanaugh  Still exist in last version..


https://www.typescriptlang.org/play?ts=4.5.0-beta#code/C4TwDgpgBAglC8UDeAoK6qkgfgFxQHIYCAaNDAQ3wAoBjfJKAE3wDsBXAWwCMIAnKAF8AlAgB8UAG4B7AJZMA3CkEoUWaACEEycunX4CG0rqhUodBszZdefElHpQAzsD6zWAcyGj4EmfKUVVXVYDQBhbTgAHygtJRRJCgEzGHCAbQBdKBiAJQhaaT4mAB4XN097VLDMiUQ0k1QMJtMaMCSKTicfCUbmvpMmlUGUDKA
